JsonSchema/ErrorMessages.cs

```diff
@@ -19,44 +19,63 @@ namespace Json.Schema;
-	{
-		var current = message;
-		var values = new object[parameters.Length];
-		for (var i = 0; i < parameters.Length; i++)
-		{
-			var parameter = parameters[i];
-			values[i] = JsonSerializer.Serialize(parameter.value, _jsonSerializerOptions);
-			current = current.Replace($"[[{parameter.token}]]", $"{{{i}}}");
-		}
-
-		return string.Format(current, values);
-	}
+    private static readonly ResourceManager _resourceManager = new("Json.Schema.Localization.Resources", typeof(ErrorMessages).Assembly);
+    private static JsonSerializerOptions? _serializerOptions;
+
+    /// <summary>
+    /// Gets or sets a culture to use for error messages.  Default is <see cref="CultureInfo.CurrentCulture"/>.
+    /// </summary>
+    public static CultureInfo? Culture { get; set; }
+
+    static ErrorMessages()
+    {
+        _serializerOptions = new JsonSerializerOptions
+        {
+#if NET8_0_OR_GREATER
+            TypeInfoResolver = JsonSchema.TypeInfoResolver
+#endif
+        };
+#if NET8_0_OR_GREATER
+        JsonSchemaSerializerContext.OptionsManager.TypeInfoResolverUpdated +=
+            (_, _) => _serializerOptions = new JsonSerializerOptions
+            {
+                TypeInfoResolver = JsonSchema.TypeInfoResolver
+            };
+#endif
+    }
+
+    private static string Get(CultureInfo? culture = null, [CallerMemberName] string? key = null)
+    {
+        if (key == null) throw new ArgumentNullException(nameof(key), "Cannot get a null-keyed resource");
+
+        if (key.StartsWith("Get"))
+            key = key.Substring(3);
+
+        return _resourceManager.GetString($"Error_{key}", culture ?? Culture ?? CultureInfo.CurrentCulture) ??
+               throw new KeyNotFoundException($"Could not find error message with key '{key}'");
+    }
+
+    /// <summary>
+    /// Replaces tokens in the form of `[[token]]` with a specified value, serialized as JSON.
+    /// </summary>
+    /// <param name="message">The message template.</param>
+    /// <param name="parameters">
+    /// Tuple of the token name (without brackets) and the value which will replace it.
+    /// </param>
+    /// <returns>The detokenized string.</returns>
+    [UnconditionalSuppressMessage("Trimming", "IL2026:Members annotated with 'RequiresUnreferencedCodeAttribute' require dynamic access otherwise can break functionality when trimming application code", Justification = "We guarantee that the SerializerOptions covers all the types we need for AOT scenarios.")]
+    [UnconditionalSuppressMessage("AOT", "IL3050:Members annotated with 'RequiresUnreferencedCodeAttribute' require dynamic access otherwise can break functionality when trimming application code", Justification = "We guarantee that the SerializerOptions covers all the types we need for AOT scenarios.")]
+    public static string ReplaceTokens(this string message, params (string token, object? value)[] parameters)
```

Comment:
Technically we can't suppress these.  This is available to users, and they could pass in something we haven't covered.

We could change the params to (string, JsonNode?) perhaps.  That puts the serialization burden on the user, but that also means we'd have more call-site updates.  (Search for usages of `EvaluationResults.Fail()`.)

Reply:
If we change this to RDC/RUC, does that balloon out to a bunch of call sites that need to be updated? It sounds like the right thing to do either way.

Reply:
Yes, it would.  Could you do that in another PR, please?

Reply:
Actually, to make that work we would need callers to pass in a IJsonTypeInfoResolver. I'm thinking we should add an AOT-friendly overload to handle it? Do you have an example of a caller using this with some types we didn't know about?

Reply:
Now at https://github.com/gregsdennis/json-everything/pull/642